PenLineModel.h

```diff
@@ -65,9 +68,18 @@ class CAVEWHERE_SKETCH_LIB_EXPORT PenLineModel : public QAbstractItemModel {
 
     // Utility methods to add or clear lines.
     Q_INVOKABLE int addNewLine();
-    void addLine(const PenLine& line);
     Q_INVOKABLE void addPoint(int lineIndex, PenPoint point);
-    void clear();
+
+    // This creates and pushes an undo command for the current last-line.
+    Q_INVOKABLE void finishNewLine();
+
+    Q_INVOKABLE void undo();
+    Q_INVOKABLE void redo();
+    Q_INVOKABLE void clear();
+
+    Q_INVOKABLE bool canUndo() const { return m_undoStack.canUndo(); }
+    Q_INVOKABLE bool canRedo() const { return m_undoStack.canRedo(); }
+    Q_INVOKABLE bool canClear() const { return !m_lines.empty(); }
```

Comment:
For canRedo(), canUndo(). Just return qundostack and you it directly in qml. 

https://doc.qt.io/qt-6/qundostack.html